fix(router): make string navigation resolve siblings/root intuitively

Vheissu opened the pull request:
- make `RouteContext` treat string instructions like `/foo`, `foo`, and `../foo` as expected by walking to the right router instead of blindly climbing the tree
- adjust the docs to explain the new sibling/parent/root rules and refresh the examples that used to warn about “doesn’t work”
- add regression coverage in the router load/href specs so both anchor clicks and imperative `router.load()` calls exercise the new behaviour

So I wrote some tests for this problem first knowing they would fail and then I made the tests pass. Like other changes I've made, I don't necessarily know the entire context of the router, so if I did something we do differently elsewhere, just let me know.

Also, special thanks to @Sayan751 for pointing me at the exact code that handles the context stuff so I didn't have to dig too far.

Closes #2256

Vheissu committed:
fix(router): make string navigation resolve siblings/root intuitively

- make `RouteContext` treat string instructions like `/foo`, `foo`, and `../foo` as expected by walking to the right router instead of blindly climbing the tree
- adjust the docs to explain the new sibling/parent/root rules and refresh the examples that used to warn about “doesn’t work”
- add regression coverage in the router load/href specs so both anchor clicks and imperative `router.load()` calls exercise the new behaviour

Vheissu committed:
test(router): cover sibling navigation + root fallback

oops, didn't run entire suite and missed this

Vheissu committed:
test(router): sigh

Vheissu committed:
test(router): please work

Vheissu committed:
test(router): these are passing locally

sigh

Vheissu committed:
feat(testing): add dom query helpers and testid queries

Introduces a reusable DOM querying utility for tests and wires it into the fixture API. Replaces ad-hoc querying logic with a single helper that supports selectors and data-testid queries, including strict single/at-most-one checks and SVG/HTML compatibility. Exposes new test-id querying methods (getByTestId, getAllByTestId, queryByTestId) and updates fixture typings to surface them, while aligning tests with data-testid attributes (legacy data-test markers are no longer supported).

Vheissu committed:
refactor(tests): switch to data-testid attributes

Aligns test selectors to data-testid convention by migrating attributes from data-test-* across examples, tests, and fixtures.
Enhances test harness with test-id helpers (getByTestId, etc.) and adopts data-testid lookups in tests for robustness.
Updates e2e and router tests and HTML fixtures to use the new attributes throughout.

Vheissu committed:
test(router): migrate to getbytestid helper

Refactors router resource load tests to use a dedicated test-id helper instead of DOM queries.

Vheissu committed:
test(router): adjust smoke tests for multiple viewports

Tests now consume getAllBy and pick the first au-viewport as the root, enabling robustness with multiple viewports in nested routes.

Relates to #2256

Sayan751 commented on packages/router/src/route-context.ts:
```diff
@@ -646,6 +648,199 @@ export class RouteContext {
+  while (index < value.length && value.charCodeAt(index) === 47 /* / */) {
+    index++;
+  }
+  return value.slice(index);
```
I think that this function can be rewritten as:

```ts
  return value.replace(/^\/+/, '');
```

Sayan751 commented on packages/router/src/route-context.ts:
```diff
@@ -646,6 +648,199 @@ export class RouteContext {
+  return value.slice(index);
+}
+
+function consumeLeadingDotSegments(path: string, startContext: RouteContext): { path: string; context: RouteContext; changed: boolean } {
```
I think this function needs to be simplified.

- Early termination cases (without staring a loop): if the path is `./foo` (-> `foo`)  or `.` (-> empty string).
- Reserve the loop only for the paths starting with `..`. Even in that case, you can make an early return from the loop if the path becomes empty, without `continue`ing.

Sayan751 reviewed (commented):
@Vheissu Thank you for tackling this! Here is the first round of review. I ahve sort of ignored the changes related to `data-testid` :)

Vheissu committed:
refactor(route-context): remove contextChanged usage

Removes contextChanged tracking from navigation normalisation and related interfaces. Simplifies path handling by consolidating leading slash and dot segment logic, and always returning updated instruction+context. Adds contextual resolution through recognition to infer the nearest suitable ancestor, enabling relative navigation without explicit ../ prefixes.

Vheissu commented:
@Sayan751 Thanks for the sharp feedback. I dropped the stale `contextChanged` flag in `createViewportInstructions` so we still normalize string instructions without carrying redundant state. 

`stripLeadingSlashes` now uses `value.replace(/^/+/, '')`. `consumeLeadingDotSegments` handles the simple `.` and `./` cases up front, loops only for `..`, and exits early once the path is exhausted, so `/.foo` stays literal. I also added a short doc
  comment above `resolveContextThroughRecognition` to explain why that lightweight lookup remains. It lets deeply nested components load sibling or root routes without needing `../` prefixes. 

Let me know if you’d like any other tweaks or extra coverage.

Vheissu committed:
Merge branch 'master' into feat/issue-2256

Vheissu committed:
Merge branch 'master' into feat/issue-2256

Vheissu committed:
refactor(router): use url parser for route recognition

Replaces ad-hoc string slicing with an expression-based approach, enabling correct handling of composite and nested route definitions. Reads the parser from router options and uses the parsed expression to determine the first component name. Fixes #2256

Vheissu committed:
fix(router): fix path resolution with dot segments

Refactors navigation normalization to accept and use the router's url parser, ensuring correct resolution of relative instructions after dot segments. This prevents incorrect ancestor-context matching and aligns behaviour with issue #2256.
